gammapy/scripts/image_pipe.py

```diff
@@ -175,13 +222,102 @@ def make_maps(self, radius, bkg_norm=True):
 
     def make_psf(self):
         log.info('Making PSF ...')
 
-    def make_exposure(self):
-        log.info('Making EXPOSURE ...')
+    def make_total_excess(self):
+        """Compute excess between counts and bkg map."""
+        self.maps["total_excess"] = SkyMap.empty_like(self.empty_image)
+        excess = self.maps["total_counts"].data - self.maps["total_bkg"].data
+        self.maps["total_excess"].data = excess
+
+    def make_exposure(self, obs_id, spectral_index=2.3, exposure_weighted_spectrum=False):
+        """Compute the exposure map for one observation.
+
+        Excess/exposure will give the differential flux at the energy Eref at the middle of the self.energy_band
+
+        If exposure_weighted_spectrum is true, it will give the integrated flux over the self.energy_band
+
+        Exposure is define as fallow:
+
+        .. math ::
+
+            EXPOSURE = \int A(E) \phi_{E} * time
+
+        assuming a power law for the flux :math: `\phi_{E} = \phi_{Eref}*\frac{E}{E_ref}^{spectral_index}`.
+
+        If exposure_weighted_spectrum is true, :math: `EXPOSURE = \int A(E) \phi_{E} * time dE / \int \phi_{E} dE`
+
+
+        Parameters
+        ----------
+        obs_id : int
+            Number of the observation
+        spectral_index : float
+            Assumed power-law spectral index
+        exposure_weighted_spectrum : bool
+            True if you want that the total excess / exposure gives the integrated flux
+
+        """
+        # TODO: should be re-implemented using the exposure_cube function
+        exposure = SkyMap.empty_like(self.empty_image)
+
+        # Determine offset value for each pixel of the map
+        xpix_coord_grid, ypix_coord_grid = exposure.coordinates(coord_type='pix')
+        # calculate pixel offset from center (in world coordinates)
+        coord = pixel_to_skycoord(xpix_coord_grid, ypix_coord_grid, exposure.wcs, origin=0)
+        center = SkyCoord.from_pixel(self.header["NAXIS1"] / 2., self.header["NAXIS2"] / 2., exposure.wcs)
+        offset = coord.separation(center)
+
+        obs = self.data_store.obs(obs_id=obs_id)
+        livetime = obs.observation_live_time_duration
+
+        # 2D Exposure computation on the self.energy_range and on an offset_tab
+        energy = EnergyBounds.equal_log_spacing(self.energy_band[0].value, self.energy_band[1].value, 100,
+                                                self.energy_band.unit)
+        energy_band = energy.bands
+        energy_bin = energy.log_centers
+        eref = EnergyBounds(self.energy_band).log_centers
+        spectrum = (energy_bin / eref) ** (-spectral_index)
+        aeff2d = obs.aeff
+        offset_tab = Angle(np.linspace(self.offset_band[0].value, self.offset_band[1].value, 10), self.offset_band.unit)
+        exposure_tab = np.sum(aeff2d.evaluate(offset_tab, energy_bin).to("cm2") * spectrum * energy_band, axis=1)
+        if exposure_weighted_spectrum:
```

Comment:
just one minute! I wanted to fix the doctring
Le 17 avr. 2016 à 21:25, Christoph Deil <notifications@github.com> a écrit :

> In gammapy/scripts/image_pipe.py:
> 
> > +        offset = coord.separation(center)
> > +
> > +        obs = self.data_store.obs(obs_id=obs_id)
> > +        livetime = obs.observation_live_time_duration
> > +
> > +        # 2D Exposure computation on the self.energy_range and on an offset_tab
> > +        energy = EnergyBounds.equal_log_spacing(self.energy_band[0].value, self.energy_band[1].value, 100,
> > +                                                self.energy_band.unit)
> > +        energy_band = energy.bands
> > +        energy_bin = energy.log_centers
> > +        eref = EnergyBounds(self.energy_band).log_centers
> > +        spectrum = (energy_bin / eref) ** (-spectral_index)
> > +        aeff2d = obs.aeff
> > +        offset_tab = Angle(np.linspace(self.offset_band[0].value, self.offset_band[1].value, 10), self.offset_band.unit)
> > +        exposure_tab = np.sum(aeff2d.evaluate(offset_tab, energy_bin).to("cm2") * spectrum * energy_band, axis=1)
> > +        if exposure_weighted_spectrum:
> Yes it's OK for me.
> 
> I only wanted to get rid of the duplicated code, so that we have one version to use and debug.
> 
> Merge?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
>